### PR TITLE
Use `ClassValue` for the map whose key is `Class`

### LINKED ISSUE #3138: ClassLoader leak in AnnotatedValueResolver
See: https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java#L709

The key of the static map is a `Class`, which means the map should use weak keys.

### PR DESCRIPTION
**Pull request by minwoox**
Close #3138

### CODE REVIEW

**Review by ikhoon — APPROVED** [review]
Thanks! @minwoox

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java [hunk 1]
```diff
@@ -136,7 +136,7 @@
     /**
      * An instance map for reusing converters, exception handlers and decorators.
      */
-    private static final ConcurrentMap<Class<?>, Object> instanceCache = new ConcurrentHashMap<>();
+    private static final ConcurrentMap<Class<?>, Object> instanceCache = new MapMaker().weakKeys().makeMap();
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java:139]: By the way maybe for a separate PR but I recently learned of `ClassValue`

https://docs.oracle.com/javase/7/docs/api/java/lang/ClassValue.html

It's like a map from class to value, but handled natively by Java with optimal performance.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java:139]: Thanks! I have not known that class. 😄

Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java:139]: Here's an example :) https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/c6eee706604a9edf35758d90dc974e4ee82a0f0a/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/decorator/BaseDecorator.java#L25

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java]: Fixed. 😉

**Review by trustin — APPROVED** [review]
Great work, @minwoox :bow: Thanks for introducing `ClassValue` to us, @anuraaga :+1:

**Pull request merged**